pkg/importer/http_data_source.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, fmt.Sprintf("unable to parse endpoint %q", endpoint))
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	httpReader, err := createHTTPReader(ctx, ep, accessKey, secKey, certDir)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if accessKey != "" && secKey != "" {
+		ep.User = url.UserPassword(accessKey, secKey)
+	}
+	dataProvider := &HTTPDataSource{
+		ctx:         ctx,
+		cancel:      cancel,
+		httpReader:  httpReader,
+		contentType: contentType,
+		endpoint:    ep,
+		customCA:    certDir != "",
+	}
+	// We know this is a counting reader, so no need to check.
+	countingReader := httpReader.(*util.CountingReader)
+	go dataProvider.pollProgress(countingReader, 10*time.Minute, time.Second)
+	return dataProvider, nil
+}
+
+// Info is called to get initial information about the data.
+func (dp *HTTPDataSource) Info() (ProcessingPhase, error) {
+	var err error
+	dp.readers, err = NewFormatReaders(dp.httpReader, dp.contentType)
+	if err != nil {
+		klog.Errorf("Error creating readers: %v", err)
+		return Error, err
+	}
+	// The readers now contain all the information needed to determine if we can stream directly or if we need scratch space to download
+	// the file to, before converting.
+	if dp.readers.Qemu && !dp.readers.Archived && !dp.customCA {
+		// We can pass straight to conversion from the endpoint. No scratch required.
+		dp.url = dp.endpoint
+		return Convert, nil
+	}
+	if dp.contentType == cdiv1.DataVolumeArchive {
+		return TransferTarget, nil
+	}
+	return Transfer, nil
+}
+
+// Transfer is called to transfer the data from the source to a scratch location.
+func (dp *HTTPDataSource) Transfer(path string) (ProcessingPhase, error) {
+	if dp.contentType == cdiv1.DataVolumeKubeVirt {
+		if util.GetAvailableSpace(path) <= int64(0) {
+			//Need scratch space but none provided.
+			return Error, ErrRequiresScratchSpace
+		}
+		file := filepath.Join(path, tempFile)
+		err := StreamDataToFile(dp.readers.TopReader(), file)
+		if err != nil {
+			return Error, err
+		}
+		// If we successfully wrote to the file, then the parse will succeed.
+		dp.url, _ = url.Parse(file)
+		return Process, nil
+	} else if dp.contentType == cdiv1.DataVolumeArchive {
+		if err := util.UnArchiveTar(dp.readers.TopReader(), path); err != nil {
+			return Error, errors.Wrap(err, "unable to untar files from endpoint")
+		}
+		dp.url = nil
+		return Complete, nil
+	}
+	return Error, errors.Errorf("Unknown content type: %s", dp.contentType)
+}
+
+// Process is called to do any special processing before giving the URI to the data back to the processor
+func (dp *HTTPDataSource) Process() (ProcessingPhase, error) {
+	return Convert, nil
+}
+
+// GetURL returns the URI that the data processor can use when converting the data.
+func (dp *HTTPDataSource) GetURL() *url.URL {
+	return dp.url
+}
+
+// Close all readers.
+func (dp *HTTPDataSource) Close() error {
+	var err error
+	if dp.readers != nil {
```

Comment:
Need to close dp.httpReader as well, right?

Reply:
No because the httpReader is part of the reader stack

```go
dp.readers, err = NewFormatReaders(dp.httpReader, dp.contentType)
```
Its the bottom of the stack after that call.